app/src/main/java/com/example/drawingApp/ColorPickerUtility.kt

```diff
@@ -0,0 +1,222 @@
+        val previousBox = view.findViewById<TextView>(R.id.colorTwo)
+        val previousColor = onColorTextForceUpdate().previousColor
+        previousBox.setBackgroundColor(previousColor.toArgb())
+        //used to prevent dragging when clicking on the gradient or the slider
+        val onViewTouch = {
+            sheet.behavior.isDraggable = false
+        }
+        val onViewLift = {
+            sheet.behavior.isDraggable = true
+        }
+        //used to set the color of the active box
+        val setActiveBox: (FloatArray) -> Unit = { hsv ->
+            val intColor = Color.HSVToColor(hsv)
+            val red = Color.red(intColor)
+            val green = Color.green(intColor)
+            val blue = Color.blue(intColor)
+
+            hView.forceText(hsv[0].toString())
+            sView.forceText((hsv[1] * 100).toString())
+            vView.forceText((hsv[2] * 100).toString())
+            rView.forceText(red.toString())
+            gView.forceText(green.toString())
+            bView.forceText(blue.toString())
+            chView.forceText(Integer.toHexString(Color.rgb(red, green, blue)).drop(2))
+
+            square.hue = hsv[0]
+            square.sat = hsv[1]
+            square.valu = hsv[2]
+            square.invalidate()
+
+            slider.sliderSetHue(hsv[0])
+
+            activeBox.setBackgroundColor(intColor)
+        }
+        //called whenever text needs to be updated
+        val updateForceViews = {
+            val pack = onColorTextForceUpdate()
+            setActiveBox(floatArrayOf(pack.hue, pack.saturation, pack.value))
+        }
+        previousBox.setOnClickListener {
+            val hsv = FloatArray(3)
+            Color.colorToHSV(previousColor.toArgb(), hsv)
+            //used to set the color of the active box to the previous color
+            setActiveBox(hsv)
+        }
+        submit.setOnClickListener {
+            onSubmission(
+                Color.HSVToColor(
+                    floatArrayOf(
+                        slider.hue,
+                        square.sat,
+                        square.valu
+                    )
+                ).toColor()
+            )
+            sheet.dismiss()
+        }
+        cancel.setOnClickListener {
+            sheet.dismiss()
+        }
+
+        slider.onHueChange = {
+            square.hue = it
+            square.invalidate()
+            onHueUpdate(it)
+            updateForceViews()
+        }
+        square.onSatOrValChange = { saturation, value ->
+            onSatOrValUpdate(saturation, value)
+            square.sat = saturation
+            square.valu = value
+            updateForceViews()
+        }
+        square.onDown = onViewTouch
+        square.onUp = onViewLift
+        slider.onDown = onViewTouch
+        slider.onUp = onViewLift
+
+        hView.addTextChangedListener {
+            val passedInt =
+                hView.text.toString().toFloatOrNull()?.let { string ->
```

Comment:
I put this in a color Util function and just had it return itself as a boolean, then I used an if statement to validate that the input was good. We should probably ask Nic how we should handle incorrect inputs in the edit text, some applications I have used for drawing before would ignore the letters and update just based on the colors in there.

Most likely the answer will be to not update until the error is corrected, but you never know for sure.

To be clear, I don't think you need to change this if it is working for you

Reply:
I think this implementation is fine, because it will only take something in the valid range of input because everything gets sanitized to be in range, and if its not able to be taken as a float its taken as null and in that case nothing happens.